### PR TITLE
Interpretation data model1

### PR DESCRIPTION
**Pull request by sbreiff**
New schemas:
* Note abstract schema
* NoteInterpretation and NoteStandard items
* VariantSampleList item used to contain variants sent to interpretation space

Other comments:
- notes have soft links to variant-samples/variants/genes while these items have a hard LinkTo to note. One of these can be changed to a calculated property, depending on how the Interpretation Space frontend and backend is implemented.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 [tool.poetry]
 # Note: Various modules refer to this system as "encoded", not "cgap-portal".
 name = "encoded"
-version = "4.3.3"
+version = "4.4.3"
```
Comment (willronchetti) [inline pyproject.toml:4]: Minor, but `4.4.0` is what you want here.

File: src/encoded/schemas/note.json [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        "variant_id": {
+            "title": "Variant ID",
+            "description": "identifier for the variant in the knowledgebase this note is related to",
+            "type": "string"
+        },
+        "gene_id": {
+            "title": "Gene ID",
+            "description": "identifier for the gene in the knowledgebase this note is related to",
+            "type": "string"
+        }
```
Comment (willronchetti) [inline src/encoded/schemas/note.json:30]: Instead of doing it this way, why not have this under a single array field (and maybe take advantage of nested):

```
"associated_items": {
            "type": "array",
            "items": {
                "type": "object",
                "properties": {
                    "item_type": {
                        "type": "string"
                    },
                    "item_identifier": {
                        "type": "string"
                    }
                }
            }
        }
```

I think all the important functionality can be solved with the above which won't tie down the design of the item so much. You might be able to get away with just an array of string identifiers as well. I would strongly advise against baking in something very specific like the original unless you are very sure you will only want to associate a single variant/variant sample/gene per note item. It might be that this is our use case now but one could imagine a situation where you want more. For example if you want to port these notes to fourfront, a generic structure will make it trivial.

File: src/encoded/schemas/note_standard.json [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        { "$ref": "mixins.json#/tags" },
+        { "$ref": "note.json#/properties" }
+    ],
+    "properties": {
+        "schema_version": {
+            "default": "1"
+        },
+        "status": {
+            "default": "draft"
+        }
```
Comment (willronchetti) [inline src/encoded/schemas/note_standard.json:27]: Is this different than the mixin status? We should consider whether we want to manage "status" through the generic one or something that is item type specific - if so it should be defined in the generic note item.

File: src/encoded/schemas/report.json [hunk 4]
```diff
@@ -28,6 +28,27 @@
...
+            "title": "Variant Samples",
+            "description": "The variant samples included in this report",
+            "type": "array",
+            "items": {
+                "title": "Variant Sample",
+                "description": "A variant sample included in this report",
+                "type": "string",
+                "linkTo": "VariantSample"
+            }
+        },
```
Comment (willronchetti) [inline src/encoded/schemas/report.json:42]: You created `VariantSampleList`, why not use it here?

File: src/encoded/types/note.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    properties={
+        'title': 'Standard Notes',
+        'description': 'Listing of Standard Notes',
+    })
+class NoteStandard(Note):
+    """NoteStandard class."""
+
+    item_type = 'note_standard'
+    schema = load_schema('encoded:schemas/note_standard.json')
+    embedded_list = Note.embedded_list
```
Comment (willronchetti) [inline src/encoded/types/note.py:38]: This is unnecessary - you will already inherit an empty embedded list from the subclassing I believe. Same comment below.

File: src/encoded/types/note.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+    @calculated_property(schema={
+        "title": "Display Title",
+        "description": "Individual's Identifier",
```
Comment (willronchetti) [inline src/encoded/types/note.py:42]: Not sure `Individual` fits well here, guessing this part was just copied from elsewhere. Same comment below.

File: src/encoded/schemas/variant.json [hunk 7]
```diff
@@ -4160,6 +4163,10 @@
             "number_step": "any",
             "order": 27,
             "grouping": "Splice Predictors"
+        },
+        "interpretations.classification": {
+            "title": "Classification",
+            "order": 100
```
Comment (willronchetti) [inline src/encoded/schemas/variant.json:4169]: We should consider these `order` values - might be this is very important and should be emphasized first thing.

File: src/encoded/types/variant.py [hunk 8]
```diff
@@ -386,6 +408,20 @@ def extract_chrom_from_variant(v):
...
+    properties={
+        'title': 'Variant Sample Lists',
+        'description': 'Collection of all variant sample lists'
+    })
+class VariantSampleList(Item):
+    """ VariantSampleList class """
+
+    item_type = 'variant_sample_list'
+    schema = load_schema('encoded:schemas/variant_sample_list.json')
+    embedded_list = []
```
Comment (willronchetti) [inline src/encoded/types/variant.py:422]: Might consider embedding some things here depending on the use case you have in mind for this list.

**Review by willronchetti — COMMENTED** [review]
Generally I think this looks great - I've posed some questions that I think are appropriate that we can discuss further as needed. There is no question that more testing is needed so if you need help with that feel free to let myself or Kent know and we can help.

File: src/encoded/schemas/note_standard.json (hunk 3 above)
Comment (sbreiff) [reply src/encoded/schemas/note_standard.json:27]: This should be taken out, thanks for pointing it out. In the end I'm meaning to just use the status mixin.

File: pyproject.toml (hunk 1 above)
Comment (sbreiff) [reply pyproject.toml:4]: right, i guess i just changed the middle number without thinking about the patch number. i'll fix that.

File: src/encoded/schemas/note.json (hunk 2 above)
Comment (sbreiff) [reply src/encoded/schemas/note.json:30]: I could use an array field like you mention above. I guess the downside would be that it's harder to restrict a note to a single VariantSample + single Variant + single Gene, but probably the front-end could be constructed so that a Note added to a particular VariantSample could only be linked to that VS's Variant and Gene.

File: src/encoded/schemas/report.json (hunk 4 above)
Comment (sbreiff) [reply src/encoded/schemas/report.json:42]: I could do that - I wasn't sure if only a few VariantSamples might be selected for the report out of the list, but I can get clarification on that.

File: src/encoded/schemas/note.json (hunk 2 above)
Comment (willronchetti) [reply src/encoded/schemas/note.json:30]: How imperative is that restriction? If there is not UI/UX for easily doing so I find it hard to imagine people would create multiple links in ways we don't want. It feels like that should be a property of the API, not the schema, but ultimately I'm not sure it matters a ton. @alexkb0009 any thoughts?

Comment (sbreiff) [reply src/encoded/schemas/note.json:30]: I agree that it would make sense for it to be a property of the API rather than the schema so I'll change it.

File: src/encoded/schemas/variant.json (hunk 7 above)
Comment (sbreiff) [reply src/encoded/schemas/variant.json:4169]: I think my thought was that in the filtering tab which is meant to be pre-interpretation we wouldn't necessarily want things from Interpretation Notes showing up on a few out of many items, especially if interpretation hasn't been done yet and it might just be a blank column, but I'll look more into how the "order" is used and what the ideal value might be.

File: src/encoded/schemas/note_interpretation.json [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+        { "$ref": "mixins.json#/uuid" },
+        { "$ref": "mixins.json#/status" },
+        { "$ref": "mixins.json#/aliases" },
+        { "$ref": "mixins.json#/attribution" },
+        { "$ref": "mixins.json#/submitted" },
+        { "$ref": "mixins.json#/modified" },
+        { "$ref": "mixins.json#/tags" },
+        { "$ref": "note.json#/properties" }
+    ],
+    "properties": {
```
Comment (alexkb0009) [inline src/encoded/schemas/note_interpretation.json:21]: Should/could we add "title" field here for e.g. these greyed out titles that might be auto-populated by "Variant Macro" dropdowns? (Such as Overview, Gnomad, ..)

Comment (sbreiff) [reply src/encoded/schemas/note_interpretation.json:21]: We could, but I think the macros are going to be dealt with at a later stage, so I would postpone this for now.

File: src/encoded/schemas/note_interpretation.json [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+            "type": "string",
+            "enum": [
+                "benign",
+                "likely benign",
+                "VUS",
+                "likely pathogenic",
+                "pathogenic"
+            ]
+        },
+        "acmg_guidelines": {
```
Comment (alexkb0009) [inline src/encoded/schemas/note_interpretation.json:48]: Will we be using this to infer the colored markers in Annotation Space and Interpretation Space? If so, it would make things simpler for us in terms of having "not yet saved" state in UI for Interpretation Space (since Interpretation Notes would be the things changing, and not some other fields in addition)

File: src/encoded/tests/test_types_variant_sample.py [hunk 11]
```diff
@@ -1,58 +1,125 @@
 import pytest
+import webtest
 import requests  # XXX: C4-211
 from .variant_fixtures import VARIANT_SAMPLE_URL
 
 
 pytestmark = [pytest.mark.working, pytest.mark.schema]
 
+variant_uuid = "f6aef055-4c88-4a3e-a306-d37a71535d8b"
```
Comment (alexkb0009) [inline src/encoded/tests/test_types_variant_sample.py:9]: No longer needed if we get it from bgm_variant["@id"] fixture?

File: src/encoded/tests/datafixtures.py [hunk 12]
```diff
@@ -1,25 +1,9 @@
 import pytest
+import webtest
 
 from uuid import uuid4
 
 
-ORDER = [
```
Comment (alexkb0009) [inline src/encoded/tests/datafixtures.py]: This removal was merged in from master. Assuming it was just moved elsewhere in an unrelated commit.

File: src/encoded/tests/datafixtures.py [hunk 13]
```diff
@@ -32,6 +16,15 @@ def error(self, msg):
         print('ERROR: ' + msg)
 
 
+def remote_user_testapp(app, remote_user):
```
Comment (alexkb0009) [inline src/encoded/tests/datafixtures.py:19]: Moved here from test_permissions.py

File: src/encoded/tests/datafixtures.py [hunk 14]
```diff
@@ -156,6 +149,48 @@ def bgm_user(testapp, institution, bgm_project):
     return testapp.get(res.location).json
 
 
+@pytest.fixture
+def bgm_user_testapp(bgm_user, app, external_tx, zsa_savepoints):
```
Comment (alexkb0009) [inline src/encoded/tests/datafixtures.py:153]: Moved here from test_permissions.py

File: src/encoded/tests/datafixtures.py [hunk 15]
```diff
@@ -156,6 +149,48 @@ def bgm_user(testapp, institution, bgm_project):
...
+    '''TODO: maybe bring in more permissions-related fixtures into here'''
+    return remote_user_testapp(app, bgm_user['uuid'])
+
+@pytest.fixture
+def bgm_variant(bgm_user_testapp, bgm_project, institution):
```
Comment (alexkb0009) [inline src/encoded/tests/datafixtures.py:158]: New fixture, used in test_types_variant_sample.py; could be moved there instead.

File: src/encoded/tests/datafixtures.py [hunk 16]
```diff
@@ -156,6 +149,48 @@ def bgm_user(testapp, institution, bgm_project):
...
+            "hg19_hgvsg": "NC_000001.11:g.12185956del"
+        }
+        ],
+        "CHROM": "1",
+        "topmed_an": 125568
+    }
+    return bgm_user_testapp.post_json('/variants', item).json['@graph'][0]
+
+@pytest.fixture
+def bgm_test_variant_sample(bgm_variant, institution, bgm_project):
```
Comment (alexkb0009) [inline src/encoded/tests/datafixtures.py:183]: Copy of 'test_variant_sample' fixture, but using BGM data/fixtures (variant, project, institution).

File: src/encoded/tests/test_types_variant_sample.py [hunk 17]
```diff
@@ -1,58 +1,123 @@
 import pytest
+import webtest
 import requests  # XXX: C4-211
 from .variant_fixtures import VARIANT_SAMPLE_URL
 
 
 pytestmark = [pytest.mark.working, pytest.mark.schema]
 
 
+# Copied from test_permissions.py
```
Comment (alexkb0009) [inline src/encoded/tests/test_types_variant_sample.py:10]: Forgot to delete these. Will do in subsequent commit.

File: src/encoded/tests/test_types_variant_sample.py [hunk 18]
```diff
@@ -1,58 +1,123 @@
...
+                # "userid" : ... # <- This should be auto-filled by server, we'll test it .. sometime.
+            }
+        ]
     }
-    resp = es_testapp.patch_json(vsl['@id'], patch, status=200).json['@graph'][0]
-    assert resp['variant_samples'][0] == vs['@id']
+    resp = bgm_user_testapp.patch_json(vsl['@id'], patch, status=200).json['@graph'][0]
+
+    assert resp['variant_samples'][0]["variant_sample_item"] == vs['@id']
+    assert len(resp['variant_samples'][0]["date_selected"]) > 10 # Check that datetime is populated
```
Comment (alexkb0009) [inline src/encoded/tests/test_types_variant_sample.py:115]: I need to change these to test 2nd item in array, will update in next commit.

Comment (alexkb0009) [reply src/encoded/tests/test_types_variant_sample.py:115]: Also adding `assert resp['variant_samples'][1]["userid"] == bgm_user["@id"]`

File: src/encoded/tests/test_types_variant_sample.py [hunk 19]
```diff
@@ -1,58 +1,123 @@
...
+def bgm_test_variant_sample2(bgm_test_variant_sample):
+    # IS NOT pre-POSTed into DB.
+    bgm_test_variant_sample_copy = bgm_test_variant_sample.copy()
+    bgm_test_variant_sample_copy['file'] = 'other-file-name'
+    return bgm_test_variant_sample_copy
+
+
 @pytest.mark.integrated  # uses s3
-def test_bam_snapshot_download(workbook, es_testapp, test_variant_sample):
+def test_bam_snapshot_download(workbook, bgm_user_testapp, bgm_test_variant_sample):
```
Comment (alexkb0009) [inline src/encoded/tests/test_types_variant_sample.py:67]: Will try to remove these `workbook` params from all tests in this file in next commit.

File: src/encoded/tests/datafixtures.py (hunk 12 above)
Comment (alexkb0009) [reply src/encoded/tests/datafixtures.py]: Hmm.. couple of new test failures after merging in master.. I guessing we need to put in these new types (note, variantsamplelist, etc) into wherever 'ORDER' went perhaps..

Comment (alexkb0009) [reply src/encoded/tests/datafixtures.py]: All set (has been moved to conftest_settings.py)
